File: src/axe/mods.py
"""Mod freshness: declared (axe.toml) vs local (ACF) vs remote (Workshop API)."""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from typing import Literal

from axe.acf import AcfFile, parse_acf
from axe.context import Context
from axe.errors import AxeError
from axe.workshop import FetchLike, WorkshopItem, get_published_file_details

FreshnessState = Literal["current", "stale", "missing_local", "missing_remote", "unmanaged"]


@dataclass(frozen=True)
class ModFreshness:
    id: int
    state: FreshnessState
    title: str | None
    local_manifest: int | None
    latest_manifest: int | None
    local_time_updated: int | None
    latest_time_updated: int | None


@dataclass(frozen=True)
class FreshnessReport:
    total: int
    current: int
    stale: int
    missing_local: int
    missing_remote: int
    unmanaged: int
    items: list[ModFreshness]


@dataclass(frozen=True)
class ModsStatus:
    declared: int
    current: int
    stale: int
    missing_local: int
    missing_remote: int


def empty_mods_status() -> ModsStatus:
    return ModsStatus(declared=0, current=0, stale=0, missing_local=0, missing_remote=0)


def check_mod_freshness(
    declared: Sequence[int],
    local: AcfFile,
    remote: Sequence[WorkshopItem],
) -> FreshnessReport:
    remote_by_id = {item.published_file_id: item for item in remote}
    declared_set = set(declared)
    items: list[ModFreshness] = []
    current = stale = missing_local = missing_remote = unmanaged = 0

    for mod_id in declared:
        local_entry = local.mods.get(mod_id)
        remote_entry = remote_by_id.get(mod_id)

        if local_entry is None:
            items.append(
                ModFreshness(
                    id=mod_id,
                    state="missing_local",
                    title=(remote_entry.title if remote_entry else None),
                    local_manifest=None,
                    latest_manifest=None,
                    local_time_updated=None,
                    latest_time_updated=(remote_entry.time_updated if remote_entry else None),
                )
            )
            missing_local += 1
            continue

        if remote_entry and remote_entry.result != 1:
            items.append(
                ModFreshness(
                    id=mod_id,
                    state="missing_remote",
                    title=remote_entry.title or None,
                    local_manifest=local_entry.manifest,
                    latest_manifest=local_entry.latest_manifest,
                    local_time_updated=local_entry.time_updated,
                    latest_time_updated=local_entry.latest_time_updated,
                )
            )
            missing_remote += 1
            continue

        remote_time = (
            remote_entry.time_updated if remote_entry else local_entry.latest_time_updated
        )
        remote_title = remote_entry.title if remote_entry else None

        is_current = (
            local_entry.time_updated >= remote_time
            and local_entry.manifest == local_entry.latest_manifest
        )
        items.append(
            ModFreshness(
                id=mod_id,
                state="current" if is_current else "stale",
                title=remote_title,
                local_manifest=local_entry.manifest,
                latest_manifest=local_entry.latest_manifest,
                local_time_updated=local_entry.time_updated,
                latest_time_updated=remote_time,
            )
        )
        if is_current:
            current += 1
        else:
            stale += 1

    for mod_id, local_entry in local.mods.items():
        if mod_id in declared_set:
            continue
        items.append(
            ModFreshness(
                id=mod_id,
                state="unmanaged",
                title=None,
                local_manifest=local_entry.manifest,
                latest_manifest=local_entry.latest_manifest,
                local_time_updated=local_entry.time_updated,
                latest_time_updated=local_entry.latest_time_updated,
            )
        )
        unmanaged += 1

    return FreshnessReport(
        total=len(declared),
        current=current,
        stale=stale,
        missing_local=missing_local,
        missing_remote=missing_remote,
        unmanaged=unmanaged,
        items=items,
    )


def has_drift(report: FreshnessReport) -> bool:
    return report.stale > 0 or report.missing_local > 0 or report.missing_remote > 0


@dataclass(frozen=True)
class RunModsCheckResult:
    report: FreshnessReport
    warnings: list[str]


def run_mods_check(
    ctx: Context,
    *,
    fetch: FetchLike | None = None,
) -> RunModsCheckResult:
    local = _read_acf(ctx.layout.workshop_acf)
    declared = list(ctx.config.mods.ids)
    warnings: list[str] = []
    remote: list[WorkshopItem] = []
    if declared:
        try:
            remote = get_published_file_details(declared, fetch=fetch)
        except AxeError as e:
            if e.kind == "workshop_api":
                warnings.append(
                    f"workshop API unreachable: {e.message}; using ACF latest_* only"
                )
            else:
                raise
    report = check_mod_freshness(declared, local, remote)
    return RunModsCheckResult(report=report, warnings=warnings)


def read_mods_status(ctx: Context) -> tuple[ModsStatus, list[str]]:
    """Compose ModsStatus + warnings from the freshness report."""
    result = run_mods_check(ctx)
    r = result.report
    return (
        ModsStatus(
            declared=r.total,
            current=r.current,
            stale=r.stale,
            missing_local=r.missing_local,
            missing_remote=r.missing_remote,
        ),
        result.warnings,
    )


def _read_acf(path) -> AcfFile:  # noqa: ANN001 — Path
    try:
        text = path.read_text()
    except FileNotFoundError:
        return AcfFile(mods={})
    except OSError as e:
        raise AxeError("filesystem", f"reading {path}: {e}") from e
    return parse_acf(text)

File: src/core/mods.ts
import { readFile } from 'node:fs/promises';
import type { AcfFile, ManifestId, WorkshopId } from './acf.ts';
import { parseAcf } from './acf.ts';
import { AxeError } from './errors.ts';
import type { Layout } from './layout.ts';
import { type FetchLike, getPublishedFileDetails, type WorkshopItem } from './workshop.ts';

export type FreshnessState = 'current' | 'stale' | 'missing_local' | 'missing_remote' | 'unmanaged';

export type ModFreshness = {
  id: WorkshopId;
  state: FreshnessState;
  title: string | null;
  local_manifest: ManifestId | null;
  latest_manifest: ManifestId | null;
  local_time_updated: number | null;
  latest_time_updated: number | null;
};

export type FreshnessReport = {
  total: number; // declared count
  current: number;
  stale: number;
  missing_local: number;
  missing_remote: number;
  unmanaged: number;
  items: ModFreshness[]; // declared in axe.toml order, then unmanaged
};

/**
 * Pure freshness comparison. Takes the declared mod IDs from `axe.toml`,
 * the locally-installed ACF map, and the remote Workshop API response;
 * returns a typed report.
 *
 * - declared && in ACF && remote knows it && manifests/times match -> 'current'
 * - declared && in ACF && remote knows it && upstream newer        -> 'stale'
 * - declared && NOT in ACF                                          -> 'missing_local'
 * - declared && in ACF && remote DOESN'T know it (result != 1)      -> 'missing_remote'
 * - in ACF && NOT declared                                          -> 'unmanaged'
 *
 * If `remote` is empty (e.g. API was unreachable and the caller fell back),
 * we still produce a report — but per-id state uses the local ACF's
 * `latest_*` fields (no remote context, titles null).
 */
export function checkModFreshness(
  declared: readonly WorkshopId[],
  local: AcfFile,
  remote: readonly WorkshopItem[],
): FreshnessReport {
  const remoteById = new Map<bigint, WorkshopItem>();
  for (const item of remote) remoteById.set(item.published_file_id, item);

  const declaredSet = new Set<bigint>(declared);
  const items: ModFreshness[] = [];
  const counts = { current: 0, stale: 0, missing_local: 0, missing_remote: 0, unmanaged: 0 };

  for (const id of declared) {
    const localEntry = local.mods.get(id);
    const remoteEntry = remoteById.get(id);

    if (!localEntry) {
      items.push(makeMissingLocal(id, remoteEntry));
      counts.missing_local++;
      continue;
    }

    // Have local. Decide using remote if available; else fall back to ACF latest_*.
    if (remoteEntry && remoteEntry.result !== 1) {
      items.push({
        id,
        state: 'missing_remote',
        title: remoteEntry.title || null,
        local_manifest: localEntry.manifest,
        latest_manifest: localEntry.latest_manifest,
        local_time_updated: localEntry.time_updated,
        latest_time_updated: localEntry.latest_time_updated,
      });
      counts.missing_remote++;
      continue;
    }

    const remoteTimeUpdated = remoteEntry?.time_updated ?? localEntry.latest_time_updated;
    const remoteTitle = remoteEntry?.title ?? null;

    const isCurrent =
      localEntry.time_updated >= remoteTimeUpdated &&
      localEntry.manifest === localEntry.latest_manifest;

    items.push({
      id,
      state: isCurrent ? 'current' : 'stale',
      title: remoteTitle,
      local_manifest: localEntry.manifest,
      latest_manifest: localEntry.latest_manifest,
      local_time_updated: localEntry.time_updated,
      latest_time_updated: remoteTimeUpdated,
    });
    if (isCurrent) counts.current++;
    else counts.stale++;
  }

  // Unmanaged: in ACF but not declared.
  for (const [id, localEntry] of local.mods) {
    if (declaredSet.has(id)) continue;
    items.push({
      id,
      state: 'unmanaged',
      title: null,
      local_manifest: localEntry.manifest,
      latest_manifest: localEntry.latest_manifest,
      local_time_updated: localEntry.time_updated,
      latest_time_updated: localEntry.latest_time_updated,
    });
    counts.unmanaged++;
  }

  return {
    total: declared.length,
    ...counts,
    items,
  };
}

function makeMissingLocal(id: WorkshopId, remoteEntry: WorkshopItem | undefined): ModFreshness {
  return {
    id,
    state: 'missing_local',
    title: remoteEntry?.title ?? null,
    local_manifest: null,
    latest_manifest: null,
    local_time_updated: null,
    latest_time_updated: remoteEntry?.time_updated ?? null,
  };
}

export type RunModsCheckOptions = {
  fetchImpl?: FetchLike;
  signal?: AbortSignal;
};

export type RunModsCheckResult = {
  report: FreshnessReport;
  warnings: string[];
};

/**
 * Orchestrates a read-only mod-freshness check.
 *
 * Reads the layout's workshop ACF (tolerates ENOENT), queries the Workshop
 * API for declared IDs (catches `workshop_api` errors and falls back to
 * ACF-only with a warning), runs `checkModFreshness`, and returns the
 * report plus any accumulated warnings.
 *
 * Pure-ish: only IO is the network and the filesystem read. No console
 * output, no `process.exitCode` setting. Caller decides what to do with
 * the report (render, exit-code, etc.).
 */
export async function runModsCheck(
  layout: Layout,
  declared: readonly WorkshopId[],
  options: RunModsCheckOptions = {},
): Promise<RunModsCheckResult> {
  const local = await readAcf(layout.workshop_acf);

  const warnings: string[] = [];
  let remote: WorkshopItem[] = [];
  if (declared.length > 0) {
    try {
      remote = await getPublishedFileDetails(declared, options);
    } catch (e) {
      if (e instanceof AxeError && e.kind === 'workshop_api') {
        warnings.push(`workshop API unreachable: ${e.message}; using ACF latest_* only`);
      } else {
        throw e;
      }
    }
  }

  const report = checkModFreshness(declared, local, remote);
  return { report, warnings };
}

/**
 * Pure policy: returns true if the freshness report shows actionable drift.
 *
 * `unmanaged` does NOT trigger drift — installed-but-not-declared is
 * interesting but not actionable; agents shouldn't treat it as "needs sync."
 */
export function hasDrift(report: FreshnessReport): boolean {
  return report.stale > 0 || report.missing_local > 0 || report.missing_remote > 0;
}

async function readAcf(path: string): Promise<AcfFile> {
  let text: string;
  try {
    text = await readFile(path, 'utf8');
  } catch (e) {
    const code = (e as NodeJS.ErrnoException).code;
    if (code === 'ENOENT') {
      return { mods: new Map() };
    }
    throw new AxeError('filesystem', `reading ${path}: ${(e as Error).message}`);
  }
  return parseAcf(text);
}
